swim/member.go

```diff
@@ -70,46 +70,46 @@ func shuffle(members []*Member) []*Member {
 	return newMembers
 }
 
+// nonLocalOverride returns wether a change should be applied to the member.
+// This function assumes that the address of the member and the change are
+// equal.
 func (m *Member) nonLocalOverride(change Change) bool {
-	return m.aliveOverride(change) ||
-		m.suspectOverride(change) ||
-		m.faultyOverride(change) ||
-		m.leaveOverride(change)
-}
-
-func (m *Member) aliveOverride(change Change) bool {
-	return change.Status == Alive && change.Incarnation > m.Incarnation
-}
-
-func (m *Member) faultyOverride(change Change) bool {
-	return change.Status == Faulty &&
-		((m.Status == Suspect && change.Incarnation >= m.Incarnation) ||
-			(m.Status == Faulty && change.Incarnation > m.Incarnation) ||
-			(m.Status == Alive && change.Incarnation >= m.Incarnation))
-}
+	// change is younger than current member
+	if change.Incarnation > m.Incarnation {
+		return true
+	}
 
-func (m *Member) leaveOverride(change Change) bool {
-	return change.Status == Leave &&
-		m.Status != Leave && change.Incarnation >= m.Incarnation
-}
+	// change is older than current member
+	if change.Incarnation < m.Incarnation {
+		return false
+	}
 
-func (m *Member) suspectOverride(change Change) bool {
-	return change.Status == Suspect &&
-		((m.Status == Suspect && change.Incarnation > m.Incarnation) ||
-			(m.Status == Faulty && change.Incarnation > m.Incarnation) ||
-			(m.Status == Alive && change.Incarnation >= m.Incarnation))
+	// If the incarnation numbers are equal, we look at the state to
+	// determine wether the change overrides this member.
+	return statePrecedence(change.Status) > statePrecedence(m.Status)
 }
 
+// localOverride returns whether a change should be applied to to the member
 func (m *Member) localOverride(local string, change Change) bool {
-	return m.localSuspectOverride(local, change) || m.localFaultyOverride(local, change)
-}
-
-func (m *Member) localFaultyOverride(local string, change Change) bool {
-	return m.Address == local && change.Status == Faulty
-}
-
-func (m *Member) localSuspectOverride(local string, change Change) bool {
-	return m.Address == local && change.Status == Suspect
+	if m.Address != local {
+		return false
+	}
+	return change.Status == Faulty || change.Status == Suspect
+}
+
+func statePrecedence(s string) int {
```

Comment:
In the end I think it will be better to use this:

```go
type State int

const (
    Alive State = iota
    Suspect
    Faulty
    Leave
)
```

But lets keep it like this for now.